src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -382,24 +370,35 @@ private TSource[] LazyToArray()
             {
                 Debug.Assert(GetCount(onlyIfCheap: true) == -1);
 
-                var builder = new LargeArrayBuilder<TSource>(initialize: true);
+                var builder = new SparseArrayBuilder<TSource>(initialize: true);
 
-                for (SingleLinkedNode<TSource> node = _prepended; node != null; node = node.Linked)
+                if (_prepended != null)
                 {
-                    builder.Add(node.Item);
+                    builder.Reserve(_prepended.Count);
                 }
 
                 builder.AddRange(_source);
 
                 if (_appended != null)
                 {
-                    foreach (TSource item in _appended.ToArray())
-                    {
-                        builder.Add(item);
-                    }
+                    builder.Reserve(_appended.Count);
```

Comment:
The general pattern is

```cs
var builder = new SparseArrayBuilder<T>(initialize: true);

...

if (CanPredetermineSizeOfSomeItems)
{
    // Semantically equivalent to builder.AddRange(
    // Enumerable.Repeat(default(T), PredeterminedSize)),
    // except we don't add anything to the buffer.
    builder.Reserve(PredeterminedSize);
}

...

// Creates an array with a bunch of zero-initialized
// gaps in it where `Reserve` has been called
T[] array = builder.ToArray();

// Get the list of locations where gaps were added.
ArrayBuilder<Marker> markers = builder.Markers;

for (int i = 0; i < markers.Count; i++)
{
    // Each Marker tells us the index of the gap in
    // the array, as well as how wide the gap is.
    Marker marker = markers[i];

    // Copy the corresponding collection/nodes/etc.
    // to this gap
    ...
}
```

This is not the best example because we don't have to use the markers (we know exactly where the gaps are here, at the start & at the end), but you can see the above code structure in `Concat` & `SelectMany`.